src/middlewared/middlewared/plugins/api_key.py

```diff
@@ -0,0 +1,145 @@
+from middlewared.service_exception import MatchNotFound
+import middlewared.sqlalchemy as sa
+
+
+class APIKeyModel(sa.Model):
+    __tablename__ = "account_api_key"
+
+    id = sa.Column(sa.Integer(), primary_key=True)
+    name = sa.Column(sa.String(200))
+    key = sa.Column(sa.Text())
+
+
+class ApiKeyService(CRUDService):
+
+    class Config:
+        namespace = "api_key"
+        datastore = "account.api_key"
+        datastore_extend = "api_key.item_extend"
+
+    @private
+    async def item_extend(self, item):
+        item.pop("key")
+        return item
+
+    @accepts(
+        Dict(
+            "api_key_create",
+            Str("name", required=True, empty=False),
+            register=True
+        )
+    )
+    async def do_create(self, data):
+        """
+        Creates API Key.
+
+        `name` is a user-readable name for key.
+        """
+        await self._validate("api_key_create", data)
+
+        key = self._generate()
+        data["key"] = pbkdf2_sha256.encrypt(key)
+
+        data["id"] = await self.middleware.call(
+            "datastore.insert",
+            self._config.datastore,
+            data
+        )
+
+        return self._serve(data, key)
+
+    @accepts(
+        Int("id", required=True),
+        Patch(
+            "api_key_create",
+            "api_key_update",
+            ("add", Bool("reset")),
+            ("attr", {"update": True}),
+        )
+    )
+    async def do_update(self, id, data):
+        """
+        Update API Key `id`.
+
+        Specify `reset: true` to reset this API Key.
+        """
+        reset = data.pop("reset", False)
+
+        old = await self._get_instance(id)
+        new = old.copy()
+
+        new.update(data)
+
+        await self._validate("api_key_update", new, id)
+
+        key = None
+        if reset:
+            key = self._generate()
+            new["key"] = pbkdf2_sha256.encrypt(key)
+
+        await self.middleware.call(
+            "datastore.update",
+            self._config.datastore,
+            id,
+            new,
+        )
+
+        return self._serve(await self._get_instance(id), key)
+
+    @accepts(
+        Int("id")
+    )
+    async def do_delete(self, id):
+        """
+        Delete API Key `id`.
+        """
+        response = await self.middleware.call(
+            "datastore.delete",
+            self._config.datastore,
+            id
+        )
+
+        return response
+
+    @private
+    async def authenticate(self, key):
+        try:
+            key_id, key = key.split("-", 1)
+            key_id = int(key_id)
+        except ValueError:
+            return None
+
+        try:
+            db_key = await self.middleware.call("datastore.query", "account.api_key", [("id", "=", key_id)],
+                                                {"get": True})
+        except MatchNotFound:
+            return None
+
+        if not pbkdf2_sha256.verify(key, db_key["key"]):
```

Comment:
Is this lightweight enough to not block the event loop?

Reply:
```
[desktop] ~> python3 -m timeit -s 'from passlib.hash import pbkdf2_sha256; s = "*" * 4096; h = pbkdf2_sha256.encrypt("*" * 64)' 'pbkdf2_sha256.verify(s, h)'   
100 loops, best of 3: 15.7 msec per loop
```